File: 3534. Path Existence Queries in a Graph II.cpp
class Solution {
public:
    vector<int> pathExistenceQueries(
        int n,
        vector<int>& nums,
        int maxDiff,
        vector<vector<int>>& queries
    ) {

        // Store (value, original index)
        vector<pair<int, int>> sortedNodes;

        sortedNodes.reserve(n);

        for (int index = 0; index < n; index++) {
            sortedNodes.push_back(
                {nums[index], index}
            );
        }

        sort(
            sortedNodes.begin(),
            sortedNodes.end()
        );

        // Position of each original index
        vector<int> position(n);

        // Sorted values
        vector<int> sortedValues(n);

        for (int index = 0; index < n; index++) {

            sortedValues[index] =
                sortedNodes[index].first;

            position[
                sortedNodes[index].second
            ] = index;
        }

        // Number of levels for Binary Lifting
        int maxPower = 1;

        while ((1 << maxPower) <= n) {
            maxPower++;
        }

        vector<vector<int>> jump(
            maxPower,
            vector<int>(n)
        );

        int rightPointer = 0;

        // Build first jump table
        for (int leftPointer = 0;
             leftPointer < n;
             leftPointer++) {

            if (rightPointer < leftPointer) {
                rightPointer = leftPointer;
            }

            while (
                rightPointer + 1 < n &&
                sortedValues[rightPointer + 1] -
                    sortedValues[leftPointer]
                    <= maxDiff
            ) {
                rightPointer++;
            }

            jump[0][leftPointer] =
                rightPointer;
        }

        // Binary Lifting table
        for (int power = 1;
             power < maxPower;
             power++) {

            for (int index = 0;
                 index < n;
                 index++) {

                jump[power][index] =
                    jump[power - 1][
                        jump[power - 1][index]
                    ];
            }
        }

        vector<int> answer;

        answer.reserve(
            queries.size()
        );

        for (const auto& query : queries) {

            int left =
                position[query[0]];

            int right =
                position[query[1]];

            if (left > right) {
                swap(left, right);
            }

            if (left == right) {

                answer.push_back(0);

                continue;
            }

            int current = left;

            int minimumSteps = 0;

            // Binary Lifting
            for (int power = maxPower - 1;
                 power >= 0;
                 power--) {

                if (
                    jump[power][current] <
                    right
                ) {

                    current =
                        jump[power][current];

                    minimumSteps +=
                        (1 << power);
                }
            }

            if (jump[0][current] >= right) {

                answer.push_back(
                    minimumSteps + 1
                );

            } else {

                answer.push_back(-1);
            }
        }

        return answer;
    }
};

File: 3534. Path Existence Queries in a Graph II.py
class Solution:
    def pathExistenceQueries(
        self,
        n: int,
        nums: List[int],
        maxDiff: int,
        queries: List[List[int]]
    ) -> List[int]:

        nodes = []

        for index in range(n):
            nodes.append((nums[index], index))

        nodes.sort()

        position = [0] * n
        values = [0] * n

        for index in range(n):
            values[index] = nodes[index][0]
            position[nodes[index][1]] = index

        LOG = 1
        while (1 << LOG) <= n:
            LOG += 1

        jump = [[0] * n for _ in range(LOG)]

        right = 0

        for left in range(n):

            if right < left:
                right = left

            while (
                right + 1 < n and
                values[right + 1] - values[left] <= maxDiff
            ):
                right += 1

            jump[0][left] = right

        for power in range(1, LOG):
            for index in range(n):
                jump[power][index] = (
                    jump[power - 1][
                        jump[power - 1][index]
                    ]
                )

        answer = []

        for first, second in queries:

            left = position[first]
            right = position[second]

            if left > right:
                left, right = right, left

            if left == right:
                answer.append(0)
                continue

            current = left
            distance = 0

            for power in range(LOG - 1, -1, -1):

                if jump[power][current] < right:

                    current = jump[power][current]

                    distance += (1 << power)

            if jump[0][current] >= right:
                answer.append(distance + 1)
            else:
                answer.append(-1)

        return answer
